File: app/src/main/java/com/darja/flickrsearch/ui/search/SearchFragment.java
package com.darja.flickrsearch.ui.search;

import android.app.Fragment;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.darja.flickrsearch.R;
import com.darja.flickrsearch.api.FlickrApi;
import com.darja.flickrsearch.model.Photo;
import com.darja.flickrsearch.model.QueryHolder;

import java.util.List;

public class SearchFragment extends Fragment implements
    SearchFragmentView.Callbacks,
    LoadPhotosListTask.OnPhotosLoadedCallback
{

    private FlickrApi mApi;

    private SearchFragmentView mView;
    private SearchFragmentModel mModel;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_search, container, false);
    }

    @Override
    public void onViewCreated(View rootView, Bundle savedInstanceState) {
        super.onViewCreated(rootView, savedInstanceState);
        setRetainInstance(true);

        mApi = new FlickrApi(getString(R.string.api_key));

        if (rootView != null) {
            mView = new SearchFragmentView(rootView);
            mView.setCallbacks(this);
        }

        mModel = new SearchFragmentModel();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mView.setCallbacks(null);
    }

    private void loadPhotos() {
        mView.showProgress();

        LoadPhotosListTask loadPhotosListTask = new LoadPhotosListTask(mApi, this, new QueryHolder(mModel.getQuery()));
        loadPhotosListTask.execute();
    }

    @Override
    public void onNewSearch() {
        String query = mView.getQueryString();
        if (mModel.isQueryChanged(query)) {
            mModel.setQuery(query);
            loadPhotos();
        }
    }

    @Override
    public void onScrolled(int lastVisibleItem, int totalItemsCount) {
        if (lastVisibleItem > totalItemsCount - 10) {
            // todo load more photos
        }
    }

    @Override
    public void onPhotosLoaded(QueryHolder queryHolder, List<Photo> photos) {
        if (isAdded()) {
            mModel.setPhotos(photos);
            int emptyResId = 0;
            if (photos == null) {
                emptyResId = R.string.error_cannot_reach_server;
            } else if (photos.size() == 0) {
                emptyResId = R.string.nothing_found;
            }

            mView.showResults(photos, emptyResId);
            mView.hideProgress();
        }
    }
}

File: app/src/main/java/com/darja/flickrsearch/ui/search/SearchFragment.kt
package com.darja.flickrsearch.ui.search

import android.app.Fragment
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.darja.flickrsearch.R
import com.darja.flickrsearch.api.Api
import com.darja.flickrsearch.util.DPLog

class SearchFragment : Fragment(), SearchView.Callbacks {
    private lateinit var api: Api

    private lateinit var view: SearchView
    private lateinit var model: SearchModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_search, container, false)
    }

    override fun onViewCreated(rootView: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(rootView, savedInstanceState)
        retainInstance = true

        api = Api(getString(R.string.api_key))

        if (rootView != null) {
            view = SearchView(rootView)
            view.callbacks = this
        }

        model = SearchModel()
    }

    override fun onDestroy() {
        super.onDestroy()
        view.callbacks = null
    }

    private fun loadPhotos() {
        view.showProgress()
        Thread(Runnable {
            val photos = api.requestPhotos(model.query, model.page)
            photos?.forEach { DPLog.vt(Api.TAG, it.url) }
            model.photos = photos

            activity?.runOnUiThread({
                if (isAdded) {
                    view.showResults(photos)
                    view.hideProgress()
                }
            })
        }).start()
    }

    override fun onNewSearch() {
        val query = view.getQueryString()
        if (model.isQueryChanged(query)) {
            model.query = query
            loadPhotos()
        }
    }

}